### PR TITLE
Filenames in multipart requests should be encoded.

### PR DESCRIPTION
**Pull request by mikelitvinets**
Since webservers try to decode filename field it should be encoded.
Otherwise, for example for the file called `c++.txt`, pluses will
be treated as encoded spaces, and the result will be `c  .txt`.

You can find how it's done in rack:
https://github.com/rack/rack/blob/master/lib/rack/multipart/parser.rb#L152

The function `get_filename` extracts file name and [encodes](https://github.com/rack/rack/blob/master/lib/rack/multipart/parser.rb#L167) it using [URI module](https://github.com/rack/rack/blob/master/lib/rack/utils.rb#L31).

In case of `c++.txt`, pluses will be replaced with spaces because of [this](https://github.com/ruby/ruby/blob/trunk/lib/uri/common.rb#L346).

### CODE REVIEW

**Comment by mikelitvinets** [comment]
Travis shows that build 453.6 (jruby) has failed, while the others are ok. It was an error during `bundle install`, but I don't think it has something to do with my patch. Should I investigate it?

**Comment by ab** [comment]
TL;DR: I think this is actually a bug in Rack.

So I've done a little bit of research here, and I'm still trying to figure out what the correct behavior ought to be. The standards unfortunately do not appear to allow for *any* unambiguous interpretation of an input filename. And the browsers I tested each do different things, none of which comply with the standards.

[RFC2388 Section 4.4](https://tools.ietf.org/html/rfc2388#section-4.4) states that file names MAY be encoded with the [RFC 2231](https://tools.ietf.org/html/rfc2231) method. This is notably much more complex than simple percent encoding. But even so, percent encoding is *not* the same as the CGI encoding that Rack uses, which transforms `" "` into `"+"` rather than `"%20"` as strict percent encoding requires. https://en.wikipedia.org/wiki/Percent-encoding#The_application.2Fx-www-form-urlencoded_type

Compare the results:

    >> URI.unescape('foo+bar')
    => "foo+bar"
    >> URI.decode_www_form_component('foo+bar')
    => "foo bar"

The RFC describes these attributes as percent encoded, not `x-www-form-urlencoded`.

There's a new multipart standard, [RFC7578](https://tools.ietf.org/html/rfc7578) that was just approved this summer. It also says filenames MAY be percent encoded (and not x-www-form-urlencoded), but it notes that many deployed systems will send UTF-8 as is. https://tools.ietf.org/html/rfc7578#section-4.2

Regardless of the standards, my inclination for rest-client is to follow behavior established by the major browsers. I tested a few different characters that they might escape, and found these results:

Input | Chrome | Firefox | Comment
----- | ------ | ------- | ---------------------------------------------
 `é` U+00E9 | `é`    | `é`     | Both browsers transmitted U+00E9 as two bytes of raw UTF-8 
 ` ` U+0020  | ` `    | ` `     | No escaping 
 `+`  | `+`    | `+`     | No escaping 
 `=`  | `=`    | `=`     | No escaping 
 `%`  | `% `   | `%`     | No escaping, even though this is ambiguous 
 `"`  | `%22`  | `\"`    | Chrome percent encoded, while Firefox backslash escaped 


Testing with a file named `résumé %41 "1+1=2".txt` yielded these results:

**Chrome 46:**

    Content-Disposition: form-data; name="myfile"; filename="résumé %41 %221+1=2%22.txt"

**Firefox 42:**

    Content-Disposition: form-data; name="myfile"; filename="résumé %41 \"1+1=2\".txt"

Firefox performed *no* percent encoding whatsoever. It appears to have taken the extremely questionable strategy of backslash escaping the `"` and leaving the `%` unescaped.

Chrome percent encoded the `"` characters, but still didn't escape the `%`, which means the output cannot be unambiguously parsed.

Given all these inconsistent results, it doesn't seem like blanket URI encoding filenames is a good idea for rest-client at this time. Major browsers do not do this, and it seems likely to create more compatibility headaches than it solves.

**Comment by ab** [comment]
This is mentioned in https://github.com/rack/rack/issues/323, where BMorearty came to the same conclusion I did about browser behavior: https://github.com/rack/rack/issues/323#issuecomment-3609743

I'm going to close this issue here.
